doc/developer/design/20230717_alter_set_cluster.md

```diff
@@ -0,0 +1,88 @@
+# Moving source and sinks
+
+Issue: [#17417](https://github.com/MaterializeInc/materialize/issues/17417)
+
+## Context
+
+Materialize's API today allows users to specify a target cluster for sources
+and sinks on creation, but does not allow for changing the cluster at a later
+time. To change the cluster, users have to drop and recreate the object.
+
+This is problematic:
+* Sources almost always have downstream dependencies that need to be recreated,
+  too.
+* Sinks require users to recreate the topic in the downstream Kafka broker.
+
+## Goals
+
+* Users chan change the cluster on which a source or sink runs.
+
+## Non-Goals
+
+* Mixing compute and storage workload on the same cluster.
+
+## Overview
+
+We'll add support for the following declarative source and sink management
+commands to Materialize:
+
+```sql
+-- Move a source to cluster `clstr`:
+ALTER SOURCE src SET CLUSTER clstr;
+
+-- Move a sink to cluster `c2`:
+ALTER SINK snk SET CLUSTER c2;
+```
+
+Note that this is _intentionally_ lacking parens
+(`ALTER SOURCE src SET (CLUSTER = clstr)`;).
+This is for consistency with the analogous PostgreSQL command
```

Comment:
Ah, was going to comment on the lack of parens (which is the syntax I'd expect for an `ALTER`). I get the nuance based on the parameter being part of the `WITH` clause or not, it's just getting hard to keep up with all the syntax variations. As a user, I just now have to remember that these are typed distinctly:

`ALTER SOURCE SET CLUSTER = some_cluster;`

`ALTER SOURCE SET (SIZE = 3xsmall);`

`ALTER CLUSTER SET (REPLICATION FACTOR = 2);`

, even though (as a user) I'll probably be confused as to why that's the case.

Reply:
One other option to consider too:

```sql
ALTER SOURCE foo IN CLUSTER clstr;
```

Reply:
I agree that it's a bit confusing when parenthesis are need and when `WITH` is needed. Looking at those three examples I think they should be consistent in terms of parenthesis. Though I'm not sure if that means we should remove parenthesis from the last two, or add them to the first.

> The rule is that only parameters configured in the `WITH` block go inside the
parens; parameters that have dedicated syntax, like schemas, tablespaces, and
clusters, get dedicated `ALTER ... SET` syntax too.

Is this in a PostgreSQL or Materialize style guide somewhere? Looking at `ALTER INDEX`, it seems to break this rule.

https://www.postgresql.org/docs/15/sql-alterindex.html

For example the following doesn't use parenthesis:
`ALTER INDEX [ IF EXISTS ] name SET TABLESPACE tablespace_name`
While the following does use parenthesis:
`ALTER INDEX [ IF EXISTS ] name SET ( storage_parameter [= value] [, ... ] )`

The two differences between those examples that I see are:
- One uses the `SET = value` syntax while the other uses the `SET value` syntax.
- One accepts a single parameter while the other accepts multiple parameters.
Though I have no clue if either are relevant to the parenthesis.

Reply:
We should consider the syntax for changing a source's schema alongside this change. Not because we plan to do it now, but because it has the same problem. The PostgreSQL syntax would be `ALTER SOURCE ... SET SCHEMA`. If we support `ALTER SOURCE ... SET (CLUSTER = ...)`, I posit that we should also support `ALTER SOURCE ... SET (SCHEMA = ...)`. But that means we'll need to support _both_ syntaxes for setting the schema, because for PostgreSQL compatibility we'll certainly need to support the non-paren variant.

Reply:
@jkosh44 PostgreSQL has a clear philosophy on this! If the parameter is created in a `WITH (...)` block in the `CREATE` statement, then it is altered with `ALTER ... SET (...)`. If the parameter has dedicated syntax in the `CREATE` statement, than it has dedicated syntax in the `ALTER` statement—often `ALTER ... SET <PARAMETER> ...` (note no parens).